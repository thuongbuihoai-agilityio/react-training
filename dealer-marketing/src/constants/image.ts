const IMAGE = {
  url: "/images/backgrounds/home-page.png",
  alt: "This is home page",
  width: 1390,
  height: 480,
  blurDataURL: "/images/backgrounds/blur.jpg",
};

export { IMAGE };
