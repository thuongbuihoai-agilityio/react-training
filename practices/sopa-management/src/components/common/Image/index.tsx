// Libs
import { memo } from 'react';
import { Link } from 'react-router-dom';

interface ImageProps {
  url?: string;
  alt?: string;
  href?: string;
  width: number;
  height: number;
  className?: string;
  size?: string;
}

const Image = ({
  url = '',
  alt = '',
  href,
  height,
  width,
  className = '',
  size = ''
}: ImageProps) =>
  href ? (
    <Link data-testid='image-link' to={href} aria-label='link to home page'>
      <img className={className} src={url} alt={alt} width={width} height={height} sizes={size} />
    </Link>
  ) : (
    <img data-testid='image' className={className} src={url} alt={alt} />
  );

export default memo(Image);
