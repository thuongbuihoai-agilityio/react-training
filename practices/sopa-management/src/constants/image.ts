export const IMAGE = {
  whiteLogo: '/images/logos/white-logo.svg',
  blackLogo: '/images/logos/black-logo.svg',
  blackShoes: '/images/products/black-shoes.png',
  blueShoes: '/images/products/blue-shoes.png',
  grayShoes: '/images/products/gray-shoes.png',
  greenShoes: '/images/products/green-shoes.png',
  whiteShoes: '/images/products/white-shoes.png',
  blackWhiteShoes: '/images/products/black-white-shoes.png',
  banner: '/images/banner.webp',
  banner2: '/images/banner-2.webp',
  altLogo: 'This is logo sopa',
  altBlackShoes: 'This is black shoes',
  altBlueShoes: 'This is blue shoes',
  altGrayShoes: 'This is gray shoes',
  altGreenShoes: 'This is green shoes',
  altWhiteShoes: 'This is white shoes',
  altBlackWhiteShoes: 'This is back white shoes',
  altBanner: 'This is banner',
};
