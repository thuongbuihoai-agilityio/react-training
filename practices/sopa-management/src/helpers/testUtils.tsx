import { render } from '@testing-library/react';
import { createMemoryHistory } from 'history';
import {
  QueryClient,
  QueryClientProvider
} from 'react-query';
import {
  BrowserRouter,
  Router
} from 'react-router-dom';

export const renderRouterTest = (component: JSX.Element) => {
  const history = createMemoryHistory();
  return render(
    <Router location={history.location} navigator={history}>
      {component}
    </Router>
  );
};

export const renderWithRouterAndQuery = (children: JSX.Element) => {
  const queryClient = new QueryClient();

  return render(
    <BrowserRouter>
      <QueryClientProvider client={queryClient}>{children}</QueryClientProvider>
    </BrowserRouter>
  );
};
