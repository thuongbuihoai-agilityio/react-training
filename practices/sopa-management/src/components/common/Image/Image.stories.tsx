import type { Meta, StoryObj } from '@storybook/react';
import { BrowserRouter } from 'react-router-dom';

// Components
import Image from '@common/Image';

// Constants
import { IMAGE } from '@constants/image';

const meta: Meta<typeof Image> = {
  component: Image,
  decorators: [
    (Story) => (
      <BrowserRouter>
        <Story />
      </BrowserRouter>
    ),
  ],
};

export default meta;
type Story = StoryObj<typeof Image>;

export const WhiteLogo: Story = {
  args: {
    url: IMAGE.whiteLogo,
    alt: IMAGE.altLogo,
  }
};

export const BlackLogo: Story = {
  args: {
    url: IMAGE.blackLogo,
    alt: IMAGE.altLogo,
  }
};
