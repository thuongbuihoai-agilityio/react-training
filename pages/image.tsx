import Image from "next/image";
import profilePic from "../public/chocolate-tarts-cake.jpg";

const Img: React.FC = () => {
  return (
    <>
      <Image
        src={profilePic}
        alt="Picture of the author"
        // width={500} automatically provided
        // height={500} automatically provided
        // blurDataURL="data:..." automatically provided
        // placeholder="blur" // Optional blur-up while loading
      />
      <p>Welcome to my Image page!</p>
    </>
  );
}

export default Img;
