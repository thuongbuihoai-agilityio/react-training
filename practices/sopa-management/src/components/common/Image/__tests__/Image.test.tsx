// Libs
import '@testing-library/jest-dom';

// Constants
import { IMAGE } from '@constants/image';

// Helpers
import { renderRouterTest } from '@helpers/testUtils';

// Components
import Image from '@common/Image';

describe('Image component', () => {
  test('should render Image component', () => {
    const { getByTestId } = renderRouterTest(
      <Image url={IMAGE.blackLogo} href='/' width={30} height={30} />
    );
    expect(getByTestId('image-link')).toBeInTheDocument();
  });

  test('should render Image component default url', () => {
    const { getByTestId } = renderRouterTest(
      <Image url={IMAGE.blackLogo} href='/' width={30} height={30} />
    );
    expect(getByTestId('image-link')).toBeInTheDocument();
  });

  test('should render Image component default', () => {
    const { getByTestId } = renderRouterTest(
      <Image width={30} height={30} />
    );
    expect(getByTestId('image')).toBeInTheDocument();
  });

  test('matches snapshot', () => {
    const { container } = renderRouterTest(
      <Image url={IMAGE.blackLogo} href='/' width={30} height={30} />
    );
    expect(container).toMatchSnapshot();
  });
});
