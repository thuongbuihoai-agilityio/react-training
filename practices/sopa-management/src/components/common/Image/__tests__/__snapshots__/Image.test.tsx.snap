// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Image component matches snapshot 1`] = `
<div>
  <a
    aria-label="link to home page"
    data-testid="image-link"
    href="/"
  >
    <img
      alt=""
      class=""
      height="30"
      sizes=""
      src="/images/logos/black-logo.svg"
      width="30"
    />
  </a>
</div>
`;
